# PMSNA - Association rules

Clément Antheaume, Camille-Amaury Juge.

## Arules

### Weather dataset

In [30]:
!pip install -U arules
import arules as ar
from arules.utils import five_quantile_based_bins, top_bottom_10, top_5_variant_variables

Requirement already up-to-date: arules in /usr/local/lib/python3.6/dist-packages (0.0.0)


In [31]:
import pandas as pd
data = pd.read_excel('Meteo.xls')
data.head()

ID   Outlook Tempreature Humidity  Windy Class
0   1     sunny         hot     high  False     N
1   2     sunny         hot     high   True     N
2   3  overcast         hot     high  False     P
3   4      rain        mild     high  False     P
4   5      rain        cool   normal  False     P

In [32]:
rules, supp_dict = ar.create_association_rules(data.drop('ID',axis=1),max_cols=2,binning_method=five_quantile_based_bins)

Calculating all relevant supports
100%|██████████| 15/15 [00:00<00:00, 553.41it/s]

Calculating all feature level rules per variable level rule
100%|██████████| 20/20 [00:00<00:00, 349.38it/s]
Overall # of Rules:  108


In [33]:
rules.head()

antecedent               consequent  ...  lift  # of all records
0  {'Tempreature': 'cool'}   {'Humidity': 'normal'}  ...  2.00                14
1   {'Humidity': 'normal'}  {'Tempreature': 'cool'}  ...  2.00                14
2  {'Outlook': 'overcast'}   {'Tempreature': 'hot'}  ...  1.75                14
3   {'Tempreature': 'hot'}  {'Outlook': 'overcast'}  ...  1.75                14
4     {'Outlook': 'sunny'}           {'Class': 'N'}  ...  1.68                14

[5 rows x 12 columns]

In [34]:
rules[(rules['confidence'] > 0.6) & (rules['lift'] > 1) ].head()

antecedent              consequent  ...    lift  # of all records
0   {'Tempreature': 'cool'}  {'Humidity': 'normal'}  ...  2.0000                14
6            {'Class': 'N'}    {'Humidity': 'high'}  ...  1.6000                14
8   {'Outlook': 'overcast'}          {'Class': 'P'}  ...  1.5556                14
10   {'Tempreature': 'hot'}    {'Humidity': 'high'}  ...  1.5000                14
22   {'Humidity': 'normal'}          {'Class': 'P'}  ...  1.3333                14

[5 rows x 12 columns]

From the generated rules, we can extract the ones which are the rules with the highest probability and with a great lift.
For example, in the Weather dataset, we know that when the temperature is cool, the humidity is always normal. 

### Mushroom dataset

In [35]:
import pandas as pd
data = pd.read_excel('Mushroom.xlsx')
data.head()

Id  Odorant  Anneau  Chapeau bombé  Pied large  Taches  Comestible
0   1        1       0              1           1       0           1
1   2        1       0              1           1       1           1
2   3        0       0              1           1       1           1
3   4        1       0              0           1       1           0
4   5        0       0              1           1       0           1

In [36]:
rules, supp_dict = ar.create_association_rules(data.drop('Id',axis=1),max_cols=2,binning_method=five_quantile_based_bins)

Calculating all relevant supports
100%|██████████| 21/21 [00:00<00:00, 712.77it/s]

Calculating all feature level rules per variable level rule
100%|██████████| 30/30 [00:00<00:00, 385.74it/s]
Overall # of Rules:  120


In [37]:
rules.head()

antecedent           consequent  ... lift  # of all records
0         {'Anneau': '1'}  {'Comestible': '0'}  ...  2.0                24
1  {'Chapeau bombé': '0'}      {'Anneau': '1'}  ...  2.0                24
2  {'Chapeau bombé': '0'}  {'Comestible': '0'}  ...  2.0                24
3     {'Pied large': '0'}      {'Anneau': '1'}  ...  2.0                24
4     {'Pied large': '0'}  {'Comestible': '0'}  ...  2.0                24

[5 rows x 12 columns]

In [38]:
rules[(rules['confidence'] > 0.7) & (rules['lift'] > 1) ].head()

antecedent              consequent  ...    lift  # of all records
10  {'Chapeau bombé': '0'}         {'Taches': '1'}  ...  1.8000                24
11     {'Pied large': '0'}         {'Taches': '1'}  ...  1.8000                24
14     {'Pied large': '0'}        {'Odorant': '0'}  ...  1.5000                24
26         {'Taches': '0'}  {'Chapeau bombé': '1'}  ...  1.1143                24
27         {'Taches': '0'}     {'Pied large': '1'}  ...  1.1143                24

[5 rows x 12 columns]

As another example, we can from a smelly mushroom expect a large foot mushroom aswell with a 92% probability

Both the example you saw are made only with 1 antecedent and consequent. Let's now allow the association rules to contain more antecedents and consequents.

In [39]:
rules, supp_dict = ar.create_association_rules(data.drop('Id',axis=1),max_cols=4,binning_method=five_quantile_based_bins)
rules.head()

Calculating all relevant supports
100%|██████████| 56/56 [00:00<00:00, 681.21it/s]

Calculating all feature level rules per variable level rule
100%|██████████| 360/360 [00:01<00:00, 265.64it/s]
Overall # of Rules:  3160


antecedent  ... # of all records
0        {'Odorant': '1', 'Pied large': '0'}  ...               24
1  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
2  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
3  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
4            {'Odorant': '0', 'Anneau': '1'}  ...               24

[5 rows x 12 columns]

In [40]:
rules[(rules['confidence'] > 0.7) & (rules['lift'] > 1.2)].head()

antecedent  ... # of all records
0        {'Odorant': '1', 'Pied large': '0'}  ...               24
1  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
2  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
3  {'Chapeau bombé': '0', 'Pied large': '0'}  ...               24
8        {'Odorant': '1', 'Pied large': '0'}  ...               24

[5 rows x 12 columns]

We now have a lot more association rules to choose from,  but as you can see, as the dataset is small a alot of these probabilities are equal to one because there is only one example with the combination of antecedents and consequents.

In [41]:
from sklearn.datasets import load_iris
data = load_iris()
df_iris = pd.DataFrame(data.data, columns=data.feature_names)
df_iris.head()
rules, supp_dict = ar.create_association_rules(df_iris,binning_method=five_quantile_based_bins)
rules.head()

Calculating all relevant supports
100%|██████████| 15/15 [00:00<00:00, 132.28it/s]

Calculating all feature level rules per variable level rule
100%|██████████| 50/50 [00:00<00:00, 144.53it/s]
Overall # of Rules:  1386


antecedent  ... # of all records
0  {'sepal width (cm)': '(3.0 - 3.1]', 'petal len...  ...              150
1  {'sepal width (cm)': '(3.0 - 3.1]', 'petal wid...  ...              150
2  {'sepal length (cm)': '(6.5 - 7.9]', 'petal le...  ...              150
3  {'sepal length (cm)': '(6.5 - 7.9]', 'petal wi...  ...              150
4  {'petal length (cm)': '(3.9 - 4.6]', 'petal wi...  ...              150

[5 rows x 12 columns]

The same exercice on the iris dataset becomes a lot more relevant for two reasons:
* We are working with ranges of number, which allows us to cover relations with more accuracy.
* The dataset doesn't contain a dozen of lines but 10, which allows more reliable values.

Also, this kind of dataset is able to give rules not only between the features and a class but between features themselves.

In [42]:
rules[(rules['confidence'] > 0.8) & (rules['lift'] > 1.2) & (rules['ant_count'] > 4)].head()

antecedent  ... # of all records
42   {'sepal width (cm)': '(2.7 - 3.0]', 'petal wid...  ...              150
58   {'sepal width (cm)': '(2.7 - 3.0]', 'petal len...  ...              150
146  {'sepal width (cm)': '(2.7 - 3.0]', 'petal len...  ...              150
148  {'sepal width (cm)': '(2.7 - 3.0]', 'petal wid...  ...              150
150  {'sepal width (cm)': '(2.7 - 3.0]', 'petal len...  ...              150

[5 rows x 12 columns]

In [43]:
target = rules[(rules['confidence'] > 0.8) & (rules['lift'] > 1.2) & (rules['ant_count'] > 4)].iloc[0]
print(target['antecedent'], target['consequent'], target['confidence'])

{'sepal width (cm)': '(2.7 - 3.0]', 'petal width (cm)': '[0.1 - 0.2]'} {'sepal length (cm)': '[4.3 - 5.0]', 'petal length (cm)': '[1.0 - 1.5]'} 0.8333


In this example we can see that a given width for both petals and sepals of an iris, within a range, can give the approximate length of their petals and sepals with a 83% chance, which is a very correct score considering the dataset size.